File: indigo_core/src/mode.rs
use ropey::Rope;

#[derive(Clone, Debug)]
pub enum Mode {
    Normal(NormalMode),
    Insert(InsertMode),
    Command(CommandMode),
}

impl Default for Mode {
    fn default() -> Self {
        Self::Normal(NormalMode::default())
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct NormalMode {
    count: usize,
}

#[derive(Clone, Copy, Debug, Default)]
pub struct InsertMode {}

#[derive(Clone, Debug, Default)]
pub struct CommandMode {
    command: Rope,
    cursor: usize,
}

File: src/Indigo/Core/Mode.hs
module Indigo.Core.Mode
  ( Mode (..)
  , NormalMode (..)
  , GotoMode (..)
  , CommandMode (..)
  , InsertMode (..)
  )
where

import Data.Default.Class (Default (..))

data Mode
  = Normal !NormalMode
  | Goto !GotoMode
  | Command !CommandMode
  | Insert !InsertMode

instance Default Mode where
  def :: Mode
  def = Normal def

data NormalMode = NormalMode
  { count :: {-# UNPACK #-} !Word
  }

instance Default NormalMode where
  def :: NormalMode
  def =
    NormalMode
      { count = 0
      }

data GotoMode = GotoMode

instance Default GotoMode where
  def :: GotoMode
  def = GotoMode

data CommandMode = CommandMode
  { commandLine :: !()
  }

instance Default CommandMode where
  def :: CommandMode
  def =
    CommandMode
      { commandLine = ()
      }

data InsertMode = InsertMode

instance Default InsertMode where
  def :: InsertMode
  def = InsertMode
